sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -250,7 +250,8 @@ abstract class Optimizer(catalogManager: CatalogManager)
       InsertMapSortInGroupingExpressions) :+
     // This batch must be executed after the `RewriteSubquery` batch, which creates joins.
     Batch("NormalizeFloatingNumbers", Once, NormalizeFloatingNumbers) :+
-    Batch("ReplaceUpdateFieldsExpression", Once, ReplaceUpdateFieldsExpression)
+    Batch("ReplaceUpdateFieldsExpression", Once, ReplaceUpdateFieldsExpression) :+
+    Batch("RewriteCollationJoin", Once, RewriteCollationJoin)
```

Comment:
since this enlarges the query plan (adding `CollationKey`), we should do it as late as possible: the last rule in `SparkOptimizer`

Reply:
moved `RewriteCollationJoin` to `SparkOptimizer.postHocOptimizationBatches`